pkg/plugins/golang/v4/scaffolds/internal/templates/makefile.go

```diff
@@ -219,31 +211,51 @@ $(LOCALBIN):
 
 ## Tool Binaries
 KUBECTL ?= kubectl
-KUSTOMIZE ?= $(LOCALBIN)/kustomize
-CONTROLLER_GEN ?= $(LOCALBIN)/controller-gen
-ENVTEST ?= $(LOCALBIN)/setup-envtest
+KUSTOMIZE ?= $(LOCALBIN)/kustomize-$(KUSTOMIZE_VERSION)
+CONTROLLER_GEN ?= $(LOCALBIN)/controller-gen-$(CONTROLLER_TOOLS_VERSION)
+ENVTEST ?= $(LOCALBIN)/setup-envtest-$(ENVTEST_VERSION)
+GOLANGCI_LINT = $(LOCALBIN)/golangci-lint-$(GOLANGCI_LINT_VERSION)
 
 ## Tool Versions
 KUSTOMIZE_VERSION ?= {{ .KustomizeVersion }}
 CONTROLLER_TOOLS_VERSION ?= {{ .ControllerToolsVersion }}
+ENVTEST_VERSION ?= release-0.16 # branch name
+GOLANGCI_LINT_VERSION ?= v1.54.2
 
 .PHONY: kustomize
-kustomize: $(KUSTOMIZE) ## Download kustomize locally if necessary. If wrong version is installed, it will be removed before downloading.
+kustomize: $(KUSTOMIZE) ## Download kustomize locally if necessary.
 $(KUSTOMIZE): $(LOCALBIN)
-	@if test -x $(LOCALBIN)/kustomize && ! $(LOCALBIN)/kustomize version | grep -q $(KUSTOMIZE_VERSION); then \
-		echo "$(LOCALBIN)/kustomize version is not expected $(KUSTOMIZE_VERSION). Removing it before installing."; \
-		rm -rf $(LOCALBIN)/kustomize; \
-	fi
-	test -s $(LOCALBIN)/kustomize || GOBIN=$(LOCALBIN) GO111MODULE=on go install sigs.k8s.io/kustomize/kustomize/v5@$(KUSTOMIZE_VERSION)
+	$(call go-install-tool,$(KUSTOMIZE),sigs.k8s.io/kustomize/kustomize,$(KUSTOMIZE_VERSION))
 
 .PHONY: controller-gen
-controller-gen: $(CONTROLLER_GEN) ## Download controller-gen locally if necessary. If wrong version is installed, it will be overwritten.
+controller-gen: $(CONTROLLER_GEN) ## Download controller-gen locally if necessary.
 $(CONTROLLER_GEN): $(LOCALBIN)
-	test -s $(LOCALBIN)/controller-gen && $(LOCALBIN)/controller-gen --version | grep -q $(CONTROLLER_TOOLS_VERSION) || \
-	GOBIN=$(LOCALBIN) go install sigs.k8s.io/controller-tools/cmd/controller-gen@$(CONTROLLER_TOOLS_VERSION)
+	$(call go-install-tool,$(CONTROLLER_GEN),sigs.k8s.io/controller-tools/cmd/controller-gen,$(CONTROLLER_TOOLS_VERSION))
 
 .PHONY: envtest
-envtest: $(ENVTEST) ## Download envtest-setup locally if necessary.
-$(ENVTEST): $(LOCALBIN)
-	test -s $(LOCALBIN)/setup-envtest || GOBIN=$(LOCALBIN) go install sigs.k8s.io/controller-runtime/tools/setup-envtest@latest
+envtest: $(LOCALBIN) ## Download envtest-setup locally if necessary.
+	$(call go-install-tool,$(ENVTEST),sigs.k8s.io/controller-runtime/tools/setup-envtest,$(ENVTEST_VERSION))
+
+.PHONY: golangci-lint
+golangci-lint: $(LOCALBIN)
+	$(call go-install-tool,$(GOLANGCI_LINT),github.com/golangci/golangci-lint/cmd/golangci-lint,${GOLANGCI_LINT_VERSION})
+
+# go-install-tool will 'go install' any package with custom target and name of binary, if it doesn't exist
+# $1 - target path with name of binary (ideally with version)
+# $2 - package url which can be installed
+# $3 - specific version of package
+define go-install-tool
```

Comment:
would simplify here (adding `/v5` back to kustomize url)
```make
define go-install-tool
@[ -f $(1) ] || { \
set -e; \
package=$(2)@$(3) ;\
echo "Downloading $${package}" ;\
GOBIN=$(LOCALBIN) go install $${package} ;\
mv "$$(echo "$(1)" | sed "s/-$(3)$$//")" $(1) ;\
}
endef
```

Reply:
i would prefer keep go-install-tool general. Some tools can use more specific path behind version. So you will have different format for VERSION what can be confusing for user.

Reply:
here is example for ginkgo and helm:

```
CONTROLLER_GEN_VERSION:=v0.10.0
KUSTOMIZE_VERSION:=v4.5.7
GOLANGCI_LINT_VERSION:=v1.51.2
HELMIFY_VERSION:=v0.4.1
GINKGO_VERSION:=v2.9.0
# isn't possible use tag
ENVTEST_VERSION:=release-0.14
HELM_VERSION:=v3.11.1

.PHONY: ginkgo-install
ginkgo-install: ## Download ginkgo bin locally if necessary.
	$(call go-install-tool,$(GINKGO_BIN),github.com/onsi/ginkgo/v2/ginkgo,${GINKGO_VERSION})

HELM_BIN = $(GOBIN)/helm-${HELM_VERSION}
.PHONY: helm-install
helm-install:
	$(call go-install-tool,$(HELM_BIN),helm.sh/helm/v3/cmd/helm,$(HELM_VERSION))
```

Reply:
@mateusoliveira43 do you agree with current solution?

Reply:
I am little worried with the `if` about version, I think it is little hard to understand it (that is for kustomize case, right? what does `echo "$url" | rev | cut -d'/' -f 1 | rev` do? my suggestion to use `$$(echo "$(1)" | sed "s/-$(3)$//` would remove the suffix `-version` from path installation, so I think it would be still general use case by the project tools)

But we can wait for the opinion of maintainers (and I really liked the idea of centralizing the tools installation in a single function :smile:)

Reply:
yes, i agree it is little confusing. Name of binary is last part of url, this is reason why i need keep url and version separately and why i modified version in if statement. First argument is name of target binary but i need source name for mv command and i cannot trust user input.

Reply:
Since this is will be used by developers, not users of the products created using kubebuilder, I would not be so strict here (even my suggested implementation has flaws: if tool name does not have version, it fails with the error `mv: 'path' and 'path' are the same file`), focusing more in readability then validation. 

For a complex set of validations in the function, I would suggest using a programming language different from shell for that (in this case, Go), as it easier to write, read and maintain the code. But if developers follow the conventions of the Makefile (for example, to update version of a tool, just change the `_VERSION` variable instead of directly changing the tool name/path, or to add new tool, follow how the other tools are installed), they should have a nice experience with it :smile: 

Or, for a more strict way of doing (but still making a simple logic), create new binary name variables (for example, `KUSTOMIZE_NAME=kustomize`) that each path would use (for example, `KUSTOMIZE ?= $(LOCALBIN)/$(KUSTOMIZE_NAME)-$(KUSTOMIZE_VERSION)`) and the function would be like this
```Make
# go-install-tool will 'go install' any specific version of a package, if it doesn't exist
# $1 - binary name of the package
# $2 - package url which can be installed
# $3 - specific version of package

define go-install-tool
@[ -f "$(LOCALBIN)/$(1)-$(3)"] || { \
set -e; \
package=$(2)@$(3) ;\
echo "Downloading $${package}" ;\
GOBIN=$(LOCALBIN) go install $${package} ;\
mv "$(LOCALBIN)/$(1)" "$(LOCALBIN)/$(1)-$(3)" ;\
}
endef
```

Reply:
i think, we started add more complexity. i would like hear another opinion.
we can use this simple solution but we will add more lines of makefile as we will delete from shell script.

i have little worry from hide connection between KUSTOMIZE_NAME and binary file, but we can solve it with comment but it isn't very user friendly for me (developers are users for me too).  But i agree, i will need to add comment into shell.

+ We can look on this as temporary solution (maybe for 1 year or long 😄 ) because maybe golang will allow define custom name for binary: https://github.com/golang/go/issues/44469

Reply:
i look on your first suggestion again and it looks fine. Sorry i was probably little sleepy yesterday. I will test it tomorrow.

Reply:
yesss, if `go install` adds that feature, that simplifies a lot 

Yes, better to hear opinions from others!

Reply:
fixed, implemented according to your first suggestion